File: java/src/main/java/it/giannibombelli/wsc2026/giftcard/application/usecases/GiftCardIssuing.java
package it.giannibombelli.wsc2026.giftcard.application.usecases;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.application.UseCase;
import it.giannibombelli.wsc2026.giftcard.application.commands.IssueGiftCard;
import it.giannibombelli.wsc2026.giftcard.domain.events.GiftCardIssued;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCard;
import it.giannibombelli.wsc2026.giftcard.domain.ports.GiftCardRepository;

import static java.util.Objects.requireNonNull;

public final class GiftCardIssuing implements UseCase<IssueGiftCard, GiftCardIssued> {
    private final GiftCardRepository giftCardRepository;

    public GiftCardIssuing(GiftCardRepository giftCardRepository) {
        this.giftCardRepository = requireNonNull(giftCardRepository);
    }

    @Override
    public GiftCardIssued invoke(IssueGiftCard cmd) {
        Require.requireArgument(cmd, "command");

        GiftCard card = GiftCard.issue(cmd.aggregateId());

        giftCardRepository.save(card);

        return new GiftCardIssued(card.id(), card.balance());
    }
}

File: typescript/src/giftcard/application/usecases/giftCardIssuing.ts
// Caso d'uso per l'emissione di una gift card.

import type { UseCase } from '@/common/application/usecase.js';
import { GiftCard } from '../../domain/giftcard/giftCard.js';
import { GiftCardIssued, giftCardIssued } from '../../domain/events/giftCardIssued.js';
import type { GiftCardRepository } from '../../domain/ports/giftCardRepository.js';
import { IssueGiftCard } from '../commands/issueGiftCard.js';
import { requireArgument } from '@/common/utils/requireArgument.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export class GiftCardIssuing implements UseCase<IssueGiftCard, GiftCardIssued> {
  private readonly _giftCardRepository: GiftCardRepository;

  constructor(giftCardRepository: GiftCardRepository) {
    requireDependency(giftCardRepository, "giftCardRepository");
    this._giftCardRepository = giftCardRepository;
  }

  invoke(command: IssueGiftCard): GiftCardIssued {
    requireArgument(command, 'command');

    const card = GiftCard.issue(command.aggregateId);
    this._giftCardRepository.save(card);

    return giftCardIssued(card.id(), card.balance());
  }
}
